src/boot/efi/boot.c

```diff
@@ -377,91 +377,119 @@ static VOID print_status(Config *config, CHAR16 *loaded_image_path) {
         UINTN timeout;
         BOOLEAN modevar;
         _cleanup_freepool_ CHAR16 *partstr = NULL, *defaultstr = NULL;
-        UINTN x, y;
+        UINTN x_max, y_max;
+        const CHAR16 *random_seed_mode = NULL;
+        const CHAR16 *console_mode = NULL;
+        CHAR16 console_mode_value[32];
+        UINTN padding = 26;
 
         assert(config);
         assert(loaded_image_path);
 
         uefi_call_wrapper(ST->ConOut->SetAttribute, 2, ST->ConOut, config->color);
         uefi_call_wrapper(ST->ConOut->ClearScreen, 1, ST->ConOut);
 
-        Print(L"systemd-boot version:   " GIT_VERSION "\n");
-        Print(L"architecture:           " EFI_MACHINE_TYPE_NAME "\n");
-        Print(L"loaded image:           %s\n", loaded_image_path);
-        Print(L"UEFI specification:     %d.%02d\n", ST->Hdr.Revision >> 16, ST->Hdr.Revision & 0xffff);
-        Print(L"firmware vendor:        %s\n", ST->FirmwareVendor);
-        Print(L"firmware version:       %d.%02d\n", ST->FirmwareRevision >> 16, ST->FirmwareRevision & 0xffff);
-
-        Print(L"Console mode:           %d/%d\n", ST->ConOut->Mode->Mode, ST->ConOut->Mode->MaxMode - 1);
-        if (uefi_call_wrapper(ST->ConOut->QueryMode, 4, ST->ConOut, ST->ConOut->Mode->Mode, &x, &y) == EFI_SUCCESS)
-                Print(L"console size:           %d x %d\n", x, y);
-
-        Print(L"SecureBoot:             %s\n", yes_no(secure_boot_enabled()));
-
-        if (efivar_get_boolean_u8(EFI_GLOBAL_GUID, L"SetupMode", &modevar) == EFI_SUCCESS)
-                Print(L"SetupMode:              %s\n", modevar ? L"setup" : L"user");
-
-        if (shim_loaded())
-                Print(L"Shim:                   present\n");
+        void line(const CHAR16 *property, const CHAR16 *format, ...) {
```

Comment:
nested functions? noooooooo!

Reply:
just make this static. nested functions are of the devil.

also add gcc format string attribute

Reply:
(nested functions mean trampoline synthesis, which is what is so evil about them)

Reply:
I originally made this a macro. But coming from C++ I hate them with a passion.
I am using a nested function cause I don't want to pass a padding var everytime to keep the code nice(ish) looking.

If you don't mind a macro in that function, I can do that.

Reply:
macros suck, but nested functions are worse, because of the trampoline (i.e. dynamic code generation).

avoid macros for things like that, but if there's no better way, they are ok to use. nested functions are never ok in our codebase though.

(grub was actually an avid user of them, and the excorcised them in a long long procedure)